### ISSUE TITLE
Event values calculated incorrectly for a certain event on mainnet

### ISSUE THREAD

**Issue by ztane** [rsksmart/rskj#1600]
I noticed a case wherein `eth_getLogs` produces an event whose arguments differs from the event within `eth_getTransactionReceipt`. I've seen the exact same incorrect value in a very recent block in two separately seeded nodes. One of the nodes I set up myself running stock `rskj-core-3.0.0-IRIS-all.jar`. The node has been running constantly since Aug 3, and the error happens with block [3581429](https://explorer.rsk.co/block/3581429).

These were the commands I used:

```
curl http://node \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc": "2.0", "method": "eth_getLogs", "params": [{"topics": ["0x59c302f58151492b1b182fa8a09d46b77d4b6745dd0444ee0339a30e8568a980"], "address": ["0x38Ed6171e38F505d732A9e280f81738041B6E2fD"], "fromBlock": "0x36a5f5", "toBlock": "0x36a5f5"}], "id": 1}'
```

vs. 

```
curl http://node \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc": "2.0", "method": "eth_getTransactionReceipt", "params": "0x4c9d1ee786fa6b08f789f404d0804c7624d723e6af5c5832e5693a36334801ef"], "id": 45}'
```

the returned events differ:

`eth_getLogs` returned event produces

```
{
    "logIndex": "0x10",
    "blockNumber": "0x36a5f5",
    "blockHash": "0x26ddc775fc3e2740cc8275e4a8e8b487ee051bdf3b724c86ca3bb8dd1d42f5e8",
    "transactionHash": "0x4c9d1ee786fa6b08f789f404d0804c7624d723e6af5c5832e5693a36334801ef",
    "transactionIndex": "0x0",
    "address": "0x38ed6171e38f505d732a9e280f81738041b6e2fd",
    "data": "0x00000000000000000000000000000000000000000000011a743004afb7dabda300000000000000000000000000000000000000000000011afb4fa9f59f94f7bc00000000000000000000000000000000000000000000011a743004afb7dabda3000000000000000000000000000000000000000000000000871fa545e7ba3a1900000000000000000000000074d1ef7e883ec0745a6219440f985da85f1c7248",
    "topics": [
        "0x59c302f58151492b1b182fa8a09d46b77d4b6745dd0444ee0339a30e8568a980",
        "0x000000000000000000000000b5999795be0ebb5bab23144aa5fd6a02d080299f",
        "0x000000000000000000000000e700691da7b9851f2f35f8b8182c69c53ccad9db"
    ]
}
```

whereas `eth_getTransactionReceipt` produces

```
{
    "logIndex": "0x10",
    "blockNumber": "0x36a5f5",
    "blockHash": "0x26ddc775fc3e2740cc8275e4a8e8b487ee051bdf3b724c86ca3bb8dd1d42f5e8",
    "transactionHash": "0x4c9d1ee786fa6b08f789f404d0804c7624d723e6af5c5832e5693a36334801ef",
    "transactionIndex": "0x0",
    "address": "0x38ed6171e38f505d732a9e280f81738041b6e2fd",
    "data": "0x00000000000000000000000000000000000000000000011a743004afb7dabda300000000000000000000000000000000000000000000011aae30e3b8779bfc0400000000000000000000000000000000000000000000011a743004afb7dabda30000000000000000000000000000000000000000000000003a00df08bfc13e6100000000000000000000000074d1ef7e883ec0745a6219440f985da85f1c7248",
    "topics": [
        "0x59c302f58151492b1b182fa8a09d46b77d4b6745dd0444ee0339a30e8568a980",
        "0x000000000000000000000000b5999795be0ebb5bab23144aa5fd6a02d080299f",
        "0x000000000000000000000000e700691da7b9851f2f35f8b8182c69c53ccad9db"
    ]
}
```

`eth_getTransactionReceipt` return value matches the expected storage and the value shown on [RSK explorer](https://explorer.rsk.co/event/036a5f5000010c86ca3bb8dd1d42f5e8?__ctab=Log); `0x11aae30e3b8779bfc04` is the value that should be there in the event as it matches the reality and storage afterwards, but `eth_getLogs` outputs `0x11afb4fa9f59f94f7bc` instead. And the same value is returned for the same transaction on two separate nodes, even though they've been running separately for a long time before block 3581429.

**Comment by ztane** [comment]
Here as a unified diff in 32 byte chunks for easier matching:

```
--- eth_getLogs 2021-08-10 08:15:01.667024093 +0300
+++ eth_getTransactionReceipt   2021-08-10 08:14:28.838891175 +0300
@@ -1,5 +1,5 @@
 00000000000000000000000000000000000000000000011a743004afb7dabda3
-00000000000000000000000000000000000000000000011afb4fa9f59f94f7bc
+00000000000000000000000000000000000000000000011aae30e3b8779bfc04
 00000000000000000000000000000000000000000000011a743004afb7dabda3
-000000000000000000000000000000000000000000000000871fa545e7ba3a19
+0000000000000000000000000000000000000000000000003a00df08bfc13e61
 00000000000000000000000074d1ef7e883ec0745a6219440f985da85f1c7248
```

**Comment by ztane** [comment]
This might be caused by `eth_getLogs` failing to consider *all interacting transactions* in the same block

**Comment by ztane** [comment]
I might add that it is not this *one* event, but over time we've had anomalies with the events in the past, and this could suggest that they're related.

**Comment by aeidelman** [comment]
Thanks @ztane, we'll be looking into this.

**Comment by ztane** [comment]
I might add that in the past we've seen 

1. events that are emitted from failed transactions show on `getLogs`
2. events that should be there to not show up on logs
3. and only now, a verified instance of a recent event that is showing the same incorrect value on two separately initialized and separately synced nodes

but if the problem indeed is that the code is not executed in the same way when recovering the events and when executing the block in the first place, then you could get all these 3 behaviours because of that.

**Comment by aeidelman** [comment]
Thanks again @ztane , the team is starting to review this, we'll contact you here if we need some further details.

**Comment by koirikivi** [comment]
As @ztane pointed out above, we have also had issues with events not showing up in `eth_getLogs` when they should (and also showing up when they shouldn't). Here's a recent example:

eth_getLogs doesn't return logs for this transaction: https://explorer.rsk.co/tx/0xafb9f8e0dba12156dd5af25b837e523454702109a301ee59b0c0ee575df30926?__ctab=Transaction (block 3598620)

Used shell command:

```
curl  https://mainnet.sovryn.app/rpc \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getLogs", "params":[{"fromBlock":"0x36e91c","toBlock":"0x36e91c"}],"id":1}'
```
(`0x36e91c` is `3598620` in hex)

Results:
https://gist.github.com/koirikivi/90625472f91b0fbb1e3d04e72030d764#file-eth_getlogs-json

Observe that the hash of the transaction `0xafb9f8e0dba12156dd5af25b837e523454702109a301ee59b0c0ee575df30926` is nowhere to be found.

`eth_getTransactionReceipt` returns the logs for the transaction correctly, with the expected block number `0x36e91c`. Used shell command:
```
curl  https://mainnet.sovryn.app/rpc \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getTransactionReceipt", "params":["0xafb9f8e0dba12156dd5af25b837e523454702109a301ee59b0c0ee575df30926"],"id":1}'
```

Results: https://gist.github.com/koirikivi/90625472f91b0fbb1e3d04e72030d764#file-eth_gettransactionreceipt-json

I'm not sure if this is related to the same issue -- it might be, or it might be a different issue, but it's definitely an issue.

**Comment by aeidelman** [comment]
@ztane @koirikivi we're trying to reproduce this issue on some of our nodes, but we haven't been able to see this same behavior yet. We are trying a few other things, we will probably contact you to get some more information from your side in case we don't succeed.

**Comment by ztane** [comment]
```# java -version
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (build 1.8.0_292-8u292-b10-0ubuntu1~20.04-b10)
OpenJDK 64-Bit Server VM (build 25.292-b10, mixed mode)
```

on one of these nodes.

**Comment by ztane** [comment]
Ok even more information. I spun up the old node on the same server as the above. We have *another* node running on the **exact same server**, except on a **different port**, that gives out the correct logs. The difference between these nodes is that the node *"correctly working"* in this case was spun up from a more recent rsk.co snapshot. However *that* node was showing discrepancies in *another* block, which was the reason for spinning up the other node which now has the verified instance of event values calculated *incorrectly*. The one where you can see problems from my initial was restored on rsk.co snapshot of block 3530000.

Nevertheless, the fact that we've got two differently working nodes on the same server should rule out discrepancies in java versions / rskj versions as they're running with the same jar (copied to different user/directory etc) and decisively the same JRE.

**Comment by koirikivi** [comment]
@aeidelman in the interest of debugging this further, I created this script:

https://gist.github.com/koirikivi/dcd775f7290dbf21dbadfe2ec1222944

It requires Python 3.6 or later but no external libraries.

The script fetches all logs returned by `eth_getLogs` between given block numbers, and then fetches each transaction with `eth_getTransactionReceipt`. After that, it makes sure that:
- The `data` fields of logs with same `transactionHash` and `logIndex` are the same
- If a log is found from the `eth_getLogs` response, it's found from the `eth_getTransactionReceipt` (and vice versa). Note that this still won't detect logs that are missing from `eth_getLogs`.
- Logs that are found from `eth_getLogs` are not from failed transactions

Trying the script with various RSK nodes, it seems that these inconsistencies are rather common and can be found with ease.

**Comment by ztane** [comment]
```
python3 check_log_inconsistencies.py --from-block 3581429 --to-block 3581429 --rpc=http://localhost:4445
```

does give 

```
Total of 7 inconsistencies detected between blocks 3581429 and 3581429.
Inconsistencies by types:
data mismatch                                 7```

and the last reported mismatch is 

```
  block: 3581429, tx: 0x4c9d1ee786fa6b08f789f404d0804c7624d723e6af5c5832e5693a36334801ef, index: 16
  type: data mismatch
  receipt data: 0x00000000000000000000000000000000000000000000011a743004afb7dabda300000000000000000000000000000000000000000000011aae30e3b8779bfc0400000000000000000000000000000000000000000000011a743004afb7dabda30000000000000000000000000000000000000000000000003a00df08bfc13e6100000000000000000000000074d1ef7e883ec0745a6219440f985da85f1c7248
  getlogs data: 0x00000000000000000000000000000000000000000000011a743004afb7dabda300000000000000000000000000000000000000000000011afb4fa9f59f94f7bc00000000000000000000000000000000000000000000011a743004afb7dabda3000000000000000000000000000000000000000000000000871fa545e7ba3a1900000000000000000000000074d1ef7e883ec0745a6219440f985da85f1c7248
```

i.e. it does find the one from my original post

**Comment by aeidelman** [comment]
@ztane do you still have the snapshot of block 3530000 you mentioned above, with which you're being able to reproduce the problem?

**Comment by ztane** [comment]
@aeidelman we do... let me find out how to share it with you alone (and not the entire world) :D

**Comment by aeidelman** [comment]
We've finally been able to reproduce it, we were not being able to see it because we were doing tests using the master branch, which already has a fix for this. We've created a separate branch containing only a patch for this issue (https://github.com/rsksmart/rskj/tree/iris-getlogs-fix), without the additional new features that are in master that are expected to be released with v3.1.0. 

We will be releasing a formal rskj version with this patch (v3.0.1) most probably next week.

Thanks @ztane and @koirikivi for reporting and helping us do the troubleshooting.
